src/as_path_parser.rs

```diff
@@ -0,0 +1,317 @@
+            Err(Error::UnexpectedEndOfBuffer)
+        } else {
+            let byte = self.buffer[self.next];
+            self.next += 1;
+            Ok(byte)
+        }
+    }
+
+    /// Advances forward four in the buffer and returns the fours bytes as a u32.
+    fn parse_u32(&mut self) -> Result<u32> {
+        let a = self.advance()?;
+        let b = self.advance()?;
+        let c = self.advance()?;
+        let d = self.advance()?;
+        Ok(u32::from_be_bytes([a, b, c, d]))
+    }
+
+    /// Returns true if buffer is exhausted, false otherwise.
+    fn done(&self) -> bool {
+        self.next == self.buffer.len()
+    }
+
+    fn parse_attributes(mut self) -> Result<Vec<u32>> {
+        let mut paths = Vec::new();
+
+        while !self.done() {
+            debug!("self.next: {}", &self.next);
+            if let Some(path) = self.parse_attribute()? {
+                // if there are no asn's in the as path
+                if path.is_empty() {
+                    warn!("Error::NoAsPathInAttributePath {:?}", &self.buffer);
+                    return Err(Error::NoAsPathInAttributePath);
+                }
+                paths.push(path);
+            }
+        }
+
+        if paths.len() > 1 {
+            // Too many asn paths in path attributes
+            warn!("Error::MultipleAsPaths {:?}", &self.buffer);
+            Err(Error::MultipleAsPaths)
+        } else if let Some(path) = paths.pop() {
+            Ok(path)
+        } else {
+            warn!("Error::NoAsPathInAttributePath{:?}", &self.buffer);
+            Err(Error::NoAsPathInAttributePath)
+        }
+    }
+
+    /// Parses all attributes. If attribute is AS_PATH, p
+    fn parse_attribute(&mut self) -> Result<Option<Vec<u32>>> {
+        let flag = self.advance()?;
+        let type_code = self.advance()?;
+        debug!("type_code: {}", type_code);
+        let mut attribute_length: u16 = self.advance()?.into();
+
+        if (flag >> 4) & 1 == 1 {
+            attribute_length <<= 8;
+            attribute_length |= self.advance()? as u16;
+        }
+        debug!("attribute_length: {}", attribute_length);
+
+        if type_code == 2 {
+            let asn_attr_position_end = &self.next + attribute_length as usize;
+
+            let asn_path = self.parse_as_path();
+
+            if asn_attr_position_end != self.next {
+                let leftover_attr_count = asn_attr_position_end - self.next;
+                for _ in 0..leftover_attr_count {
+                    self.advance()?;
+                }
+            }
+
+            asn_path
+        } else {
+            for _ in 0..attribute_length {
+                self.advance()?;
+            }
+            Ok(None)
+        }
+    }
+
+    /// Takes self and reutnrs a vec of asn
+    fn parse_as_path(&mut self) -> Result<Option<Vec<u32>>> {
+        let as_set_indicator = self.advance()?;
+
+        // Determine if asn's are listed as an unordered AS_SET (1) or an ordered AS_SEQUENCE (2)
```

Comment:
Maybe we should do something in this case too? Have you looked into what it actually means?

Reply:
This means that there are two ways AS Paths can be listed in the `AS_Paths` bgp attribute, either as an ordered set (`AS_SEQUENCE`) or and unordered set (`AS_SET`). 

I chose to disregard the unordered `AS_SET`s because the algorithm to find the bottleneck is based on the assumption that the ASNs are ordered, correct?

If not ignored, how should the unordered sets be handled?

Reply:
Agreed. Let's just log the entire line (or ip/path pair) if we find a line like that in DEBUG or something, if this doesn't make the code much uglier? Also, I guess no need to even parse asns?

Reply:
Ok so if we hit an attribute with the `AS_SET` type, you just want a log statement to stdout stating the IP and explaining it is unordered so it is left out?

Small, but should this be an `INFO` or `DEBUG` logging level? I feel like it should be `INFO`, but will do `DEBUG` if you think that is more appropriate.

Reply:
Yeah, ideally print IP, the AS set itself, and say that it was unordered so we have no idea what to do.
I don't really care which logging level it should be, so pick something.